drivers/xen/gnttab.c

```diff
@@ -193,61 +194,93 @@ static void gop_eagain_retry(int cmd, struct gnttab_map_grant_ref *gref)
 	}
 }
 
-void *gnttab_get_page(void)
+void *gnttab_get_pages(unsigned int npages)
 {
 	int ret;
 	void *page_addr;
-	struct xen_remove_from_physmap rfpm;
+	unsigned int removed;
+	xen_pfn_t gfn;
 
-	page_addr = k_aligned_alloc(XEN_PAGE_SIZE, XEN_PAGE_SIZE);
-	if (!page_addr) {
-		LOG_WRN("Failed to allocate memory for gnttab page!\n");
+	if (npages == 0) {
 		return NULL;
 	}
 
-	rfpm.domid = DOMID_SELF;
-	rfpm.gpfn = xen_virt_to_gfn(page_addr);
+	page_addr = k_aligned_alloc(XEN_PAGE_SIZE, XEN_PAGE_SIZE * npages);
+	if (!page_addr) {
+		LOG_WRN("Failed to allocate memory for gnttab %u pages!", npages);
+		return NULL;
+	}
 
 	/*
 	 * GNTTABOP_map_grant_ref will simply replace the entry in the P2M
 	 * and not release any RAM that may have been associated with
 	 * page_addr, so we release this memory before mapping.
 	 */
-	ret = HYPERVISOR_memory_op(XENMEM_remove_from_physmap, &rfpm);
+	for (removed = 0; removed < npages; removed++) {
+		gfn = xen_virt_to_gfn((char *)page_addr + (removed * XEN_PAGE_SIZE));
+
+		ret = xendom_remove_from_physmap(DOMID_SELF, gfn);
+		if (ret) {
+			break;
+		}
+	}
+
 	if (ret) {
-		LOG_WRN("Failed to remove gnttab page from physmap, ret = %d\n", ret);
+		LOG_WRN("xendom_remove_from_physmap failed: ret=%d, removed=%u, gfn=%llx",
+			ret, removed, (uint64_t)gfn);
+
+		if (removed > 0) {
+			ret = gnttab_put_pages(page_addr, removed);
+			if (ret) {
+				LOG_ERR("gnttab_put_pages failed ret=%d addr=%p", ret, page_addr);
+				k_panic();
+			}
+		}
+
 		return NULL;
 	}
 
 	return page_addr;
 }
 
-void gnttab_put_page(void *page_addr)
+int gnttab_put_pages(void *page_addr, unsigned int npages)
```

Comment:
I think we should rename `page_addr` parameter if we are using plural form now (`start_addr`, `first_page` etc.)